urla5230a98e3_Dhana_Krishnasamy_InMemoryJMS_tgz-pJ8-com_dhana_broker_demo_AppJ8/src/com/dhana/broker/impl/QueueImpl.java

```diff
@@ -0,0 +1,102 @@
+package com.dhana.broker.impl;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.Semaphore;
+
+import com.dhana.broker.Message;
+import com.dhana.broker.exceptions.DestinationClosedException;
+import com.dhana.broker.util.MessageExpiryUtil;
+
+public class QueueImpl extends AbstractDestination implements Queue {
+	private ExecutorService executor = Executors.newSingleThreadExecutor();
+	private Semaphore permits = new Semaphore(0);
+	private MessageExpiryUtil expiryUtil;
+	private static MessageDiscardPolicy defaultDiscardPolicy = MessageDiscardPolicy.DiscardLatest;
+	private static int defaultQueueSize = 20;
+	private MessageDiscardPolicy discardPolicy;
+
+	public QueueImpl(String name) {
+		this(name, defaultQueueSize, defaultDiscardPolicy);
+	}
+
+	public QueueImpl(String name, int queueSize) {
+		this(name, queueSize, defaultDiscardPolicy);
+	}
+
+	public QueueImpl(String name, int queueSize,
+			MessageDiscardPolicy discardPolicy) {
+		super(name);
+		queue = new LinkedBlockingQueue<Message>(queueSize);
+		this.discardPolicy = discardPolicy;
+		expiryUtil = new MessageExpiryUtil(queue);
+		executor.execute(new Distributor());
+	}
+
+	public boolean put(Message message) throws DestinationClosedException {
+		boolean isSuccessful = super.put(message);
+		if (!isSuccessful) {
+			if (discardPolicy == MessageDiscardPolicy.DiscardOldest) {
+				// Until successful insert
+				while (!queue.offer(message)) {
+					queue.poll();
+				}
+				expiryUtil.scheduleForExpiry(message);
+			}
+		} else {
+			expiryUtil.scheduleForExpiry(message);
+		}
+		permits.release();
+		logger.debug("Message Added to the queue:" + message.getMessage());
+		return true;
+	}
+
+	class Distributor implements Runnable {
+
+		public void run() {
+			while (true) {
+				try {
+					logger.debug("Waiting for new message");
+					permits.acquire();
+					Message msg = queue.peek();
+					if (msg == null) {
+						// May be a new subscriber was added
+						continue;
+					}
+					if (msg == poison) {
+						logger.info("Received close signal, closing queue:"
+								+ name);
+						break;
+					}
+					if (!msg.hasExpired() && !subscribers.isEmpty()) {
+						AsyncMessageConsumer subscriber = subscribers.get(0);
+						subscriber.onMessage(msg.clone());
+						logger.debug("Message sent to a client");
+						queue.remove(msg);
+					}
+				} catch (InterruptedException e) {
+					logger.debug("Interrupted exception");
+					Thread.currentThread().interrupt();
+					break;
+				}
+			}
+
+		}
+
+	}
+
+	public void addSubscriber(AsyncMessageConsumer subscriber)
+			throws DestinationClosedException {
+		super.addSubscriber(subscriber);
+		if (!queue.isEmpty())
+			permits.release(queue.size());
+	}
+
+	public void close() {
+		super.close();
+		
+		expiryUtil.cancel();
+		executor.shutdownNow();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Call `shutdown()` to reject any incoming tasks and `awaitTermination()` to wait for existing tasks to complete or until a timeout is reached. To stop the execution and processing of tasks, if necessary, call `shutdownNow()`. Check the return value of `awaitTermination()` and handle when the method times out while waiting for other threads to stop executing. See the following diffs on GitHub that make this code change: [diff1](https://github.com/glowroot/glowroot/commit/81de5a3699c59e0cf98d485c380c1034e6cd28a5#diff-c608d2c40790b0e0e5228e6647f69729caa756bbe9127169139406314b80897eL143), [diff2](https://github.com/apache/directory-kerby/commit/3ed0e7cd3b8a0c3ce29293153e4e72f422f0637f#diff-eb42f072849f520563ec2a686789a836195f18075cc222ef355c6ea710b0a360L84)

Learn more: [ExecutorService in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html)